parlai/nn/r3f.py

```diff
@@ -0,0 +1,190 @@
+            type=str,
+            default=R3F_NOISE_UNIFORM,
+            choices=[R3F_NOISE_NORMAL, R3F_NOISE_UNIFORM],
+            help='type of noises for RXF methods',
+        )
+        group.add_argument(
+            '--r3f-encoder-noise',
+            type=bool,
+            default=False,
+            help='Add noise to encoder.',
+        )
+        group.add_argument(
+            '--r3f-decoder-noise',
+            type=bool,
+            default=False,
+            help='Add noise to decoder.',
+        )
+        return argparser
+
+    def set_r3f_settings_from_opts(self, opts):
+        self.r3f_eps = opts.get('r3f_eps')
+        self.r3f_noise_type = opts.get('r3f_noise_type')
+        if self.r3f_noise_type == R3F_NOISE_NORMAL:
+            self.r3f_noise_sampler = torch.distributions.normal.Normal(
+                loc=0.0, scale=self.r3f_eps
+            )
+        elif self.r3f_noise_type == R3F_NOISE_UNIFORM:
+            self.r3f_noise_sampler = torch.distributions.uniform.Uniform(
+                low=-self.r3f_eps, high=self.r3f_eps
+            )
+        self.r3f_lambda = opts.get("r3f_lambda")
+
+        self.noise_encoder = opts.get("r3f_encoder_noise")
+        self.noise_decoder = opts.get("r3f_decoder_noise")
+
+        # Find embedding values and store locally so we don't have to find them each turn
+        self.r3f_embeddings = {}
+        self.is_generative_model = True  # assume true until we find otherwise
+        self._find_embeddings(self.model)
+
+    def compute_loss(self, batch, return_output=False):
+        if not hasattr(self, 'r3f_lambda'):
+            self.set_r3f_settings_from_opts(self.opt)
+        loss, standard_output = super().compute_loss(batch, True)
+        with R3FNoiseEmbeddingContextManager(self, self.model) as r3f:
+            noised_result = self.model(*self._model_input(batch), ys=batch.label_vec)
+            noised_scores, _, *_ = noised_result
+            standard_scores, _, *_ = standard_output
+            r3f_loss = r3f._calculate_symm_kl(noised_scores, standard_scores)
+            # get average loss per token correctly
+            notnull = batch.label_vec.ne(self.NULL_IDX)
+            target_tokens = notnull.long().sum(dim=-1)
+            r3f_loss /= target_tokens.sum()
+            # add in loss
+            loss += self.r3f_lambda * r3f_loss
+        if return_output:
+            return (loss, standard_output)
+        else:
+            return loss
+
+    def _calculate_symm_kl(self, noised_logits, input_logits):
+        return F.kl_div(
+            F.log_softmax(noised_logits, dim=-1, dtype=torch.float32),
+            F.softmax(input_logits, dim=-1, dtype=torch.float32),
+            None,
+            None,
+            "sum",
+        ) + F.kl_div(
+            F.log_softmax(input_logits, dim=-1, dtype=torch.float32),
+            F.softmax(noised_logits, dim=-1, dtype=torch.float32),
+            None,
+            None,
+            "sum",
```

Comment:
I think to implement per-token correctly, we would need to use reduction none, and then mask out the padding, and then do .sum() / num_tokens